File: src/common/flowSymbol.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Utils_js

type kind =
  | SymbolClass
  | SymbolComponent
  | SymbolConstant
  | SymbolConstructor
  | SymbolEnum
  | SymbolEnumMember
  | SymbolFile
  | SymbolFunction
  | SymbolInterface
  | SymbolMethod
  | SymbolModule
  | SymbolNamespace
  | SymbolProperty
  | SymbolTypeAlias
  | SymbolTypeParameter
  | SymbolVariable

let string_of_kind = function
  | SymbolClass -> "class"
  | SymbolComponent -> "component"
  | SymbolConstant -> "const"
  | SymbolConstructor -> "constructor"
  | SymbolEnum -> "enum"
  | SymbolEnumMember -> "enum member"
  | SymbolFile -> "file"
  | SymbolFunction -> "function"
  | SymbolInterface -> "interface"
  | SymbolMethod -> "method"
  | SymbolModule -> "module"
  | SymbolNamespace -> "namespace"
  | SymbolProperty -> "property"
  | SymbolTypeAlias -> "type alias"
  | SymbolTypeParameter -> "type parameter"
  | SymbolVariable -> "variable"

type symbol = {
  kind: kind;
  name: string;
  def_loc: ALoc.t;
}

let mk_symbol ~kind ~name ~def_loc = { kind; name; def_loc }

let mk_class_symbol ~name = mk_symbol ~kind:SymbolClass ~name

let mk_component_symbol ~name = mk_symbol ~kind:SymbolComponent ~name

let mk_constant_symbol ~name = mk_symbol ~kind:SymbolConstant ~name

let mk_enum_symbol ~name = mk_symbol ~kind:SymbolEnum ~name

let mk_module_symbol ~name =
  mk_symbol ~kind:SymbolModule ~name:(Flow_import_specifier.display_userland name)

let mk_namespace_symbol ~name = mk_symbol ~kind:SymbolNamespace ~name

let mk_type_alias_symbol ~name = mk_symbol ~kind:SymbolTypeAlias ~name

let mk_type_parameter_symbol ~name = mk_symbol ~kind:SymbolTypeParameter ~name

let mk_variable_symbol ~name = mk_symbol ~kind:SymbolVariable ~name

(* TODO: create more constructors for other kinds of symbols as needed *)

let kind_of_symbol { kind; _ } = kind

let name_of_symbol { name; _ } = name

let def_loc_of_symbol { def_loc; _ } = def_loc

let string_of_aloc aloc =
  match ALoc.source aloc with
  | None -> ""
  | Some file -> spf "%s:%s" (File_key.to_string file) (ALoc.to_string_no_source aloc)

let dump_symbol { kind; name; def_loc } =
  spf "%s: %s %s" (string_of_aloc def_loc) (string_of_kind kind) name

let dump_symbol_opt = Base.Option.value_map ~default:"no symbol" ~f:dump_symbol

File: rust_port/crates/flow_common/src/flow_symbol.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::sync::Arc;

use dupe::Dupe;
use flow_aloc::ALoc;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;

use crate::reason::string_of_aloc;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum SymbolKind {
    SymbolClass,
    SymbolComponent,
    SymbolConstant,
    SymbolConstructor,
    SymbolEnum,
    SymbolEnumMember,
    SymbolFile,
    SymbolFunction,
    SymbolInterface,
    SymbolMethod,
    SymbolModule,
    SymbolNamespace,
    SymbolProperty,
    SymbolTypeAlias,
    SymbolTypeParameter,
    SymbolVariable,
}

impl SymbolKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::SymbolClass => "class",
            Self::SymbolComponent => "component",
            Self::SymbolConstant => "const",
            Self::SymbolConstructor => "constructor",
            Self::SymbolEnum => "enum",
            Self::SymbolEnumMember => "enum member",
            Self::SymbolFile => "file",
            Self::SymbolFunction => "function",
            Self::SymbolInterface => "interface",
            Self::SymbolMethod => "method",
            Self::SymbolModule => "module",
            Self::SymbolNamespace => "namespace",
            Self::SymbolProperty => "property",
            Self::SymbolTypeAlias => "type alias",
            Self::SymbolTypeParameter => "type parameter",
            Self::SymbolVariable => "variable",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
struct SymbolInner {
    kind: SymbolKind,
    name: FlowSmolStr,
    def_loc: ALoc,
}

#[derive(Debug, Clone, Dupe, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Symbol(Arc<SymbolInner>);

impl Symbol {
    fn new(kind: SymbolKind, name: FlowSmolStr, def_loc: ALoc) -> Self {
        Symbol(Arc::new(SymbolInner {
            kind,
            name,
            def_loc,
        }))
    }

    pub fn mk_class_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolClass, name, def_loc)
    }

    pub fn mk_component_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolComponent, name, def_loc)
    }

    pub fn mk_constant_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolConstant, name, def_loc)
    }

    pub fn mk_enum_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolEnum, name, def_loc)
    }

    pub fn mk_module_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolModule, name, def_loc)
    }

    pub fn mk_namespace_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolNamespace, name, def_loc)
    }

    pub fn mk_type_alias_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolTypeAlias, name, def_loc)
    }

    pub fn mk_type_parameter_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolTypeParameter, name, def_loc)
    }

    pub fn mk_variable_symbol(name: FlowSmolStr, def_loc: ALoc) -> Self {
        Self::new(SymbolKind::SymbolVariable, name, def_loc)
    }

    pub fn kind(&self) -> SymbolKind {
        self.0.kind
    }

    pub fn name(&self) -> &FlowSmolStr {
        &self.0.name
    }

    pub fn def_loc_of_symbol(&self) -> &ALoc {
        &self.0.def_loc
    }
}

pub fn dump_symbol(symbol: &Symbol) -> String {
    let loc_str = string_of_aloc(None, &symbol.0.def_loc);
    format!("{}: {} {}", loc_str, symbol.0.kind.as_str(), symbol.0.name)
}

pub fn dump_symbol_opt(symbol: Option<&Symbol>) -> String {
    match symbol {
        Some(s) => dump_symbol(s),
        None => "no symbol".to_string(),
    }
}
